sw/device/tests/sim_dv/flash_init_test.c

```diff
@@ -0,0 +1,312 @@
+            kNumTestWords, 0) == 0);
+  CHECK_BUFFER_EQ(readback_data, kBank1Page0Data, kNumTestWords);
+  CHECK(flash_ctrl_testutils_read(
+            &flash_state, region_addresses[kAddressCreatorSecret], kPartitionId,
+            readback_data, kDifFlashCtrlPartitionTypeInfo, kNumTestWords,
+            0) == 0);
+  CHECK_BUFFER_EQ(readback_data, kCreatorSecretData, kNumTestWords);
+  CHECK(flash_ctrl_testutils_read(
+            &flash_state, region_addresses[kAddressOwnerSecret], kPartitionId,
+            readback_data, kDifFlashCtrlPartitionTypeInfo, kNumTestWords,
+            0) == 0);
+  CHECK_BUFFER_EQ(readback_data, kOwnerSecretData, kNumTestWords);
+  CHECK(flash_ctrl_testutils_read(
+            &flash_state, region_addresses[kAddressIsoPart], kPartitionId,
+            readback_data, kDifFlashCtrlPartitionTypeInfo, kNumTestWords,
+            0) == 0);
+  CHECK_BUFFER_EQ(readback_data, kIsoPartData, kNumTestWords);
+
+  read_and_check_host_if(kRegionBaseBank0Page0Index, kBank0Page0Data);
+  read_and_check_host_if(kPageSize * kRegionBaseBank1Page0Index,
+                         kBank1Page0Data);
+}
+
+static void check_readback_fail(void) {
+  uint32_t readback_data[kNumTestWords];
+  CHECK(flash_ctrl_testutils_read(
+            &flash_state, region_addresses[kAddressBank0Page0Data],
+            kPartitionId, readback_data, kDifFlashCtrlPartitionTypeData,
+            kNumTestWords, 0) != 0);
+  CHECK(flash_ctrl_testutils_read(
+            &flash_state, region_addresses[kAddressBank1Page0Data],
+            kPartitionId, readback_data, kDifFlashCtrlPartitionTypeData,
+            kNumTestWords, 0) != 0);
+  CHECK(flash_ctrl_testutils_read(
+            &flash_state, region_addresses[kAddressCreatorSecret], kPartitionId,
+            readback_data, kDifFlashCtrlPartitionTypeInfo, kNumTestWords,
+            0) != 0);
+  CHECK(flash_ctrl_testutils_read(
+            &flash_state, region_addresses[kAddressOwnerSecret], kPartitionId,
+            readback_data, kDifFlashCtrlPartitionTypeInfo, kNumTestWords,
+            0) != 0);
+  CHECK(flash_ctrl_testutils_read(
+            &flash_state, region_addresses[kAddressIsoPart], kPartitionId,
+            readback_data, kDifFlashCtrlPartitionTypeInfo, kNumTestWords,
+            0) != 0);
+}
+
+static void flash_init(void) {
+  CHECK_DIF_OK(dif_flash_ctrl_start_controller_init(&flash_state));
+  dif_flash_ctrl_status_t flash_ctrl_status;
+  while (true) {
+    CHECK_DIF_OK(dif_flash_ctrl_get_status(&flash_state, &flash_ctrl_status));
+    if (flash_ctrl_status.controller_init_wip == false) {
+      break;
+    }
+  }
+}
+
+static void check_unscrambled_init(void) {
+  setup_unscrambled_regions();
+  erase_and_write_regions();
+  flash_init();
+  check_readback_data_match();
+  test_status_set(kTestStatusInWfi);
+  wait_for_interrupt();
+}
+
+static void check_scrambled_init(void) {
+  setup_scrambled_regions();
```

Comment:
i wanted to make sure i understand this part here.
So the reason this one should fail, is because we write the regions without going through `flash_init`, as a result it's using some default key in the system.

After you do `flash_init`, the read-back should no longer match, is that the idea? 

When you do this, we should run into an integrity error (which I think you disable the tb check for), but I actually thought the host would exception as well.  Maybe there's a bug somewhere here.

Reply:
ah i forgot you're doing this read via the controller not the Host. 
should we double check to make sure it is a read error instead of just comparing to non-zero?

Reply:
Yes it uses default keys to start. I have put an additional check after the write to read back the initial write. After the init I have modified the check to look specifically for a read error.